File: day1/solution1.1.ts
import * as fs from "fs";

function calculateTotalDistance(leftList: number[], rightList: number[]): number {
    // Step 1: Sort both lists
    leftList.sort((a, b) => a - b);
    rightList.sort((a, b) => a - b);

    // Step 2: Calculate the distances and sum them up
    let totalDistance = 0;
    for (let i = 0; i < leftList.length; i++) {
        totalDistance += Math.abs(leftList[i] - rightList[i]);
    }

    return totalDistance;
}

function parseInput(filePath: string): [number[], number[]] {
    const data = fs.readFileSync(filePath, "utf-8").trim();
    const lines = data.split("\n");

    if (lines.length < 2) {
        throw new Error("Input file must contain at least two lines.");
    }

    const leftList = lines[0].split(" ").map(Number);
    const rightList = lines[1].split(" ").map(Number);

    return [leftList, rightList];
}

// Main Execution
const inputFilePath = "input.txt";
try {
    const [leftList, rightList] = parseInput(inputFilePath);
    const result = calculateTotalDistance(leftList, rightList);
    console.log("Total Distance:", result);
} catch (error) {
    console.error("Error:", error.message);
}

File: day1/solution1.1.py
def calculate_total_distance(left_list, right_list):
    """
    Calculate the total distance between two lists.
    """
    # Step 1: Sort both lists
    left_list.sort()
    right_list.sort()

    # Step 2: Calculate the total distance
    total_distance = sum(abs(l - r) for l, r in zip(left_list, right_list))
    return total_distance


def parse_input(file_path):
    """
    Parse the input file and return two lists of numbers.
    """
    with open(file_path, 'r') as file:
        lines = file.readlines()

    if len(lines) < 2:
        raise ValueError("Input file must contain at least two lines.")

    left_list = list(map(int, lines[0].strip().split()))
    right_list = list(map(int, lines[1].strip().split()))

    return left_list, right_list


if __name__ == "__main__":
    input_file_path = "input.txt"

    try:
        # Parse input and calculate the total distance
        left_list, right_list = parse_input(input_file_path)
        result = calculate_total_distance(left_list, right_list)
        print("Total Distance:", result)
    except Exception as e:
        print("Error:", str(e))